test/e2e_node/shortened_grace_period_test.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"context"
+	"github.com/onsi/ginkgo/v2"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	admissionapi "k8s.io/pod-security-admission/api"
+	"time"
+)
+
+var _ = SIGDescribe("Shortened Grace Period", func() {
+	f := framework.NewDefaultFramework("shortened-grace-period")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelBaseline
+	ginkgo.Context("When repeatedly deleting pods", func() {
+		var podClient *e2epod.PodClient
+		ginkgo.BeforeEach(func() {
+			podClient = e2epod.NewPodClient(f)
+		})
+		ginkgo.It("shorter grace period of a second command overrides the longer grace period of a first command", func() {
```

Comment:
This verifies that we exit, but it doesn't test the semantics the pod sees.

The pod should expect to see (for all container runtimes):

```
SIGTERM
SIGTERM
SIGKILL
```

We don't need to verify the KILL because other code verifies that containers are terminated, but we *should* structure this test so that if the CRI delivers more than 2 SIGINT it's an error.

I would recommend changing your bash script to be pseudocode:

```
trap INT func()
func() {
  if // first INT {
     // write "SIGINT 1"
  }
  if // second INT {
     // write "SIGINT 2"
     // sleep 5s (less than shortened grace period, long enough to detect a second int, long enough to gather logs in test)
     exit 0
  }
  if // received more than 2 INT {
    // write "SIGINT <N>"
    exit 1
  }
}
```

Then I would expect the e2e test to be doing additional things:

1. Increase the short grace period slightly so we have more room (20s is fine)
2. Stream the logs from the pod and verify you see exactly `SIGINT 1\nSIGINT 2\n` - no more, no less (shouldn't be flaky anymore, but best to test)
3. Create a test watch stream using `WatchEventSequenceVerifier` and verify that you see the container exit code for the container be zero, not any other value (1 means we sent too many sig ints, 137 means something else went wrong)

That will prevent any future regressions of runtime behavior.

Then there are two other variants we need to test:

1. Same logic that is currently in the test pod, but instead place the logic in a preStop hook - the hook should write its exit status as a file (0 or 1) in an empty dir, and the container command should read that file (missing is exit code 2, return 0 or 1 if the file matches, return exit code 3 otherwise) and return the appropriate exit code.  Most of the test can be the same
2. Create an uninterruptable preStopHook (same as what you have in the PR right now - traps SIGINT forever), keep the short grace period short (1s is fine) and verify that the pod terminates promptly.  Currently the kubelet can fall behind in test e2e and take a while to report pod cleanup status, so you can't verify it stops too fast or the test will be flaky.

Finally, we should track the total latency from the test of when we invoke DeleteSync to the time we see the pod is finally deleted and write that out as an `e2e.Logf` - eventually I will add a stress e2e test like what https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/test/e2e/node/pods.go#L210 does that takes your code and ensures that we have bounded latency on create - that's the prerequisite for us closing #113606 because if the test passes and is not flaky it means we have no code inside sync*pod that can't be interrupted.

The great news is that when you add this e2e test, we have everything we need to fix the rest of the "shorten timeout" gaps and bring kubelet status latency down.  Thank you for doing this work!

Reply:
@smarterclayton are you able to give any pointers here?